File: go/transport.go
package observability

import (
	"context"
	"errors"
	"net/http"
	"sync"
	"time"

	"github.com/SmooAI/fetch/go/fetch"
)

// Batched HTTP transport. Holds a bounded queue, flushes on a timer or when
// MaxBatchSize events are buffered, and retries a failed batch by pushing it
// back to the front of the queue. Errors are swallowed — observability must
// never throw into host code. Mirrors the TS Transport.
//
// Outbound delivery goes through github.com/SmooAI/fetch/go/fetch (SMOODEV-2026)
// so each batch POST gets the resilient stack — jittered exponential-backoff
// retries on 429/5xx + network errors, a per-request timeout, and an optional
// circuit breaker — instead of a bare net/http call. fetch handles transient
// blips within a single Flush; the transport's re-queue-on-failure handles
// longer outages across Flush cycles, so the two retry layers are complementary
// (fast in-call recovery vs. durable cross-cycle persistence), not duplicative:
// fetch aborts immediately on 4xx (RetryAbort), matching the old permanent-error
// behavior, so a persistent client error is not retried twice in a row.

const (
	defaultFlushMillis = 1000
	defaultBatchSize   = 30
	defaultQueueMax    = 250
)

// Transport is the batched event sender.
type Transport struct {
	dsn           string
	flushInterval time.Duration
	maxBatch      int
	maxQueue      int
	client        *fetch.Client

	mu       sync.Mutex
	queue    []ObservabilityEvent
	timer    *time.Timer
	inFlight bool
	closed   bool
}

// TransportOptions configures a Transport.
type TransportOptions struct {
	DSN           string
	FlushInterval time.Duration
	MaxBatchSize  int
	MaxQueueSize  int
	// HTTPClient overrides the underlying *http.Client that the resilient fetch
	// client drives (test seam). When nil, fetch's default transport is used.
	HTTPClient *http.Client
}

// NewTransport builds a Transport from options, applying defaults.
func NewTransport(opts TransportOptions) *Transport {
	flush := opts.FlushInterval
	if flush <= 0 {
		flush = defaultFlushMillis * time.Millisecond
	}
	batch := opts.MaxBatchSize
	if batch <= 0 {
		batch = defaultBatchSize
	}
	queue := opts.MaxQueueSize
	if queue <= 0 {
		queue = defaultQueueMax
	}
	return &Transport{
		dsn:           opts.DSN,
		flushInterval: flush,
		maxBatch:      batch,
		maxQueue:      queue,
		client:        buildFetchClient(opts.HTTPClient),
	}
}

// buildFetchClient assembles the resilient fetch client used for batch delivery.
// It keeps the prior 10s timeout and JSON content type, adds default retries
// (429/5xx + network, jittered backoff, aborts on 4xx), and lets a test seam
// swap the underlying *http.Client.
func buildFetchClient(httpClient *http.Client) *fetch.Client {
	retry := fetch.DefaultRetryOptions
	b := fetch.NewClientBuilder().
		WithTimeout(10 * time.Second).
		WithRetry(&retry).
		WithBaseHeaders(http.Header{"Content-Type": []string{"application/json"}})
	if httpClient != nil {
		b = b.WithHTTPClient(httpClient)
	}
	return b.Build()
}

// newTransportFromClientOptions wires a Transport from ClientOptions, used by
// the bootstrap / default init when a DSN is configured.
func newTransportFromClientOptions(opts ClientOptions) *Transport {
	flush := time.Duration(opts.FlushInterval) * time.Millisecond
	return NewTransport(TransportOptions{
		DSN:           opts.DSN,
		FlushInterval: flush,
		MaxBatchSize:  opts.MaxBatchSize,
		MaxQueueSize:  opts.MaxQueueSize,
	})
}

// Enqueue adds an event to the queue, dropping the oldest if the queue is full
// (recent events are more useful), and flushes immediately when a full batch is
// buffered.
func (t *Transport) Enqueue(event ObservabilityEvent) {
	defer recoverSilently()
	t.mu.Lock()
	if t.closed {
		t.mu.Unlock()
		return
	}
	if len(t.queue) >= t.maxQueue {
		t.queue = t.queue[1:]
	}
	t.queue = append(t.queue, event)
	full := len(t.queue) >= t.maxBatch
	if !full && t.timer == nil {
		t.timer = time.AfterFunc(t.flushInterval, func() { _ = t.Flush(context.Background()) })
	}
	t.mu.Unlock()

	if full {
		_ = t.Flush(context.Background())
	}
}

// Flush sends one batch. Safe to call concurrently — a single batch is in
// flight at a time. Returns nil on success or when there's nothing to send.
func (t *Transport) Flush(ctx context.Context) error {
	t.mu.Lock()
	if t.inFlight || len(t.queue) == 0 {
		t.clearTimerLocked()
		t.mu.Unlock()
		return nil
	}
	t.inFlight = true
	n := t.maxBatch
	if n > len(t.queue) {
		n = len(t.queue)
	}
	batch := make([]ObservabilityEvent, n)
	copy(batch, t.queue[:n])
	t.queue = t.queue[n:]
	t.clearTimerLocked()
	t.mu.Unlock()

	err := t.send(ctx, batch)

	t.mu.Lock()
	t.inFlight = false
	if err != nil {
		// Best-effort: push the batch back to the front for the next attempt.
		t.queue = append(batch, t.queue...)
		if len(t.queue) > t.maxQueue {
			t.queue = t.queue[len(t.queue)-t.maxQueue:]
		}
	}
	if len(t.queue) > 0 && t.timer == nil && !t.closed {
		t.timer = time.AfterFunc(t.flushInterval, func() { _ = t.Flush(context.Background()) })
	}
	t.mu.Unlock()
	return err
}

func (t *Transport) send(ctx context.Context, batch []ObservabilityEvent) error {
	payload := IngestPayload{Type: "error", Events: batch}
	// fetch marshals the body to JSON and applies retries/timeout/backoff. It
	// returns a typed *fetch.HTTPResponseError for non-2xx responses after
	// retries are exhausted; normalize that to httpStatusError so callers and
	// tests see the same surface as before.
	resp, err := fetch.SimplePost(ctx, t.client, t.dsn, payload, nil)
	if err != nil {
		var httpErr *fetch.HTTPResponseError
		if errors.As(err, &httpErr) {
			return &httpStatusError{status: httpErr.StatusCode}
		}
		return err
	}
	if !resp.OK {
		return &httpStatusError{status: resp.StatusCode}
	}
	return nil
}

// Close stops the flush timer and prevents further enqueues. A final Flush
// should be called by the caller before Close if draining is desired.
func (t *Transport) Close() {
	t.mu.Lock()
	t.closed = true
	t.clearTimerLocked()
	t.mu.Unlock()
}

func (t *Transport) clearTimerLocked() {
	if t.timer != nil {
		t.timer.Stop()
		t.timer = nil
	}
}

// queueSize is a test seam.
func (t *Transport) queueSize() int {
	t.mu.Lock()
	defer t.mu.Unlock()
	return len(t.queue)
}

type httpStatusError struct{ status int }

func (e *httpStatusError) Error() string {
	return "observability: ingest returned HTTP " + http.StatusText(e.status)
}

File: dotnet/src/SmooAI.Observability/Transport.cs
using System.Net.Http;
using System.Text;
using SmooAI.Fetch;

namespace SmooAI.Observability;

/// <summary>
/// Tunables for <see cref="Transport"/>. Defaults match the TS SDK.
/// </summary>
public sealed class TransportOptions
{
    /// <summary>Ingest endpoint: <c>POST /webhooks/observability/{org_id}/{token}</c>.</summary>
    public string Dsn { get; set; } = string.Empty;

    /// <summary>Flush interval in ms. Default 1000.</summary>
    public int FlushIntervalMs { get; set; } = 1000;

    /// <summary>Max events per flush batch. Default 30.</summary>
    public int MaxBatchSize { get; set; } = 30;

    /// <summary>Max events kept in memory waiting to flush. Default 250.</summary>
    public int MaxQueueSize { get; set; } = 250;

    /// <summary>Per-request timeout. Default 10s.</summary>
    public TimeSpan RequestTimeout { get; set; } = TimeSpan.FromSeconds(10);

    /// <summary>
    /// Retries applied <em>within a single flush attempt</em> by the resilient
    /// <see cref="SmooFetch"/> transport (transient exceptions + 429/5xx). This is
    /// distinct from the transport's own batch-requeue, which only kicks in once a
    /// flush has exhausted these retries and still failed. Default 2.
    /// </summary>
    public int MaxRetries { get; set; } = 2;
}

/// <summary>
/// Batched HTTP transport. Holds a small bounded queue, flushes on a timer or
/// when <see cref="TransportOptions.MaxBatchSize"/> events are buffered, and
/// retries a failed batch by pushing it back to the front of the queue. Errors
/// are swallowed — observability must never throw into the host application.
///
/// Outbound delivery goes through <see cref="SmooFetch"/> (the SmooAI resilient
/// fetch — Polly-backed retry, per-request timeout, circuit breaking) by default.
/// A caller-supplied <see cref="HttpClient"/> is honored as an escape hatch and
/// used directly (no SmooFetch resilience layer), for tests or hosts that need a
/// fully custom pipeline.
///
/// Port of the TS <c>Transport</c> (without the browser <c>sendBeacon</c> path,
/// which has no .NET analogue).
/// </summary>
public sealed class Transport : IDisposable, IAsyncDisposable
{
    private readonly TransportOptions _options;
    // Exactly one of these is set: _fetch by default, _httpClient when the caller
    // supplies a custom client via the escape hatch.
    private readonly SmooFetch? _fetch;
    private readonly HttpClient? _httpClient;
    private readonly bool _ownsHttpClient;
    private readonly Queue<ObservabilityEvent> _queue = new();
    private readonly object _gate = new();
    private readonly Timer _timer;
    private bool _inFlight;
    private bool _disposed;

    /// <summary>
    /// Create a transport. By default delivery uses an internally-managed,
    /// resilient <see cref="SmooFetch"/>. If <paramref name="httpClient"/> is
    /// supplied it is used directly as an escape hatch — the caller owns its
    /// disposal and is responsible for any resilience behavior.
    /// </summary>
    public Transport(TransportOptions options, HttpClient? httpClient = null)
    {
        _options = options ?? throw new ArgumentNullException(nameof(options));
        if (string.IsNullOrEmpty(_options.Dsn))
        {
            throw new ArgumentException("Transport requires a DSN.", nameof(options));
        }

        if (httpClient is null)
        {
            // Default path: resilient SmooFetch with the transport's timeout +
            // retry config and a circuit breaker so a hard-down ingest endpoint
            // stops hammering after repeated failures.
            _fetch = SmooFetch.Create(o =>
            {
                o.Timeout = _options.RequestTimeout;
                o.RetryPolicy = _options.MaxRetries > 0
                    ? RetryPolicy.ExponentialBackoff(_options.MaxRetries)
                    : RetryPolicy.None;
                // Stop hammering a hard-down ingest endpoint: open after 5
                // consecutive failures for 30s. A tripped breaker throws, which
                // PostAsync catches and reports as a failed delivery, so the
                // batch is requeued rather than lost.
                o.CircuitBreaker = new CircuitBreakerOptions(
                    FailureThreshold: 5,
                    OpenDuration: TimeSpan.FromSeconds(30));
            });
            _ownsHttpClient = false;
        }
        else
        {
            // Escape hatch: use the caller's client verbatim.
            _httpClient = httpClient;
            _httpClient.Timeout = _options.RequestTimeout;
            _ownsHttpClient = false;
        }

        // Disarmed timer; armed on demand when the first event is enqueued.
        _timer = new Timer(_ => _ = FlushAsync(), null, Timeout.Infinite, Timeout.Infinite);
    }

    /// <summary>
    /// Queue an event for delivery. Drops the oldest event if the queue is full
    /// (recent events are more useful). Triggers an immediate flush once the
    /// batch threshold is reached, otherwise arms the flush timer.
    /// </summary>
    public void Enqueue(ObservabilityEvent ev)
    {
        if (ev is null || _disposed)
        {
            return;
        }

        bool flushNow = false;
        lock (_gate)
        {
            if (_queue.Count >= _options.MaxQueueSize)
            {
                _queue.Dequeue();
            }
            _queue.Enqueue(ev);

            if (_queue.Count >= _options.MaxBatchSize)
            {
                flushNow = true;
            }
            else
            {
                ArmTimer();
            }
        }

        if (flushNow)
        {
            _ = FlushAsync();
        }
    }

    /// <summary>
    /// Flush up to <see cref="TransportOptions.MaxBatchSize"/> queued events.
    /// On failure the batch is restored to the front of the queue and the timer
    /// re-armed. Never throws.
    /// </summary>
    public async Task FlushAsync()
    {
        List<ObservabilityEvent> batch;
        lock (_gate)
        {
            if (_inFlight || _queue.Count == 0)
            {
                DisarmTimer();
                return;
            }
            _inFlight = true;
            batch = DequeueBatch();
            DisarmTimer();
        }

        try
        {
            var payload = new IngestPayload { Type = "error", Events = batch };
            var json = ObservabilityJson.Serialize(payload);
            var delivered = await PostAsync(json).ConfigureAwait(false);
            // Non-2xx (or transport failure) is a delivery failure: requeue for
            // the next attempt.
            if (!delivered)
            {
                RestoreBatch(batch);
            }
        }
        catch
        {
            // Network / serialization failure — best-effort requeue.
            RestoreBatch(batch);
        }
        finally
        {
            lock (_gate)
            {
                _inFlight = false;
                if (_queue.Count > 0)
                {
                    ArmTimer();
                }
            }
        }
    }

    /// <summary>
    /// POST the serialized payload via SmooFetch (default) or the escape-hatch
    /// <see cref="HttpClient"/>. Returns <c>true</c> on a 2xx, <c>false</c>
    /// otherwise. Never throws (transport exceptions are caught and reported as a
    /// failed delivery so the batch is requeued).
    /// </summary>
    private async Task<bool> PostAsync(string json)
    {
        if (_fetch is not null)
        {
            // SmooFetch serializes typed bodies itself, but we need the exact
            // ObservabilityJson wire bytes (camelCase + omit-nulls), so build the
            // request with pre-serialized content and use the low-level SendAsync
            // (which applies retry/timeout/circuit-breaking but does NOT throw on
            // non-2xx — we inspect the status ourselves).
            using var request = new HttpRequestMessage(HttpMethod.Post, _options.Dsn)
            {
                Content = new StringContent(json, Encoding.UTF8, "application/json"),
            };
            try
            {
                using var response = await _fetch.SendAsync(request).ConfigureAwait(false);
                return response.IsSuccessStatusCode;
            }
            catch
            {
                // Retries / circuit breaker exhausted — treat as a failed delivery.
                return false;
            }
        }

        // Escape hatch: raw HttpClient, no SmooFetch resilience.
        using var content = new StringContent(json, Encoding.UTF8, "application/json");
        using var raw = await _httpClient!.PostAsync(_options.Dsn, content).ConfigureAwait(false);
        return raw.IsSuccessStatusCode;
    }

    /// <summary>Current queue depth — exposed for tests.</summary>
    public int QueueSize
    {
        get
        {
            lock (_gate)
            {
                return _queue.Count;
            }
        }
    }

    private List<ObservabilityEvent> DequeueBatch()
    {
        var take = Math.Min(_options.MaxBatchSize, _queue.Count);
        var batch = new List<ObservabilityEvent>(take);
        for (var i = 0; i < take; i++)
        {
            batch.Add(_queue.Dequeue());
        }
        return batch;
    }

    private void RestoreBatch(List<ObservabilityEvent> batch)
    {
        lock (_gate)
        {
            // Rebuild the queue with the failed batch at the front, preserving order.
            var pending = _queue.ToArray();
            _queue.Clear();
            foreach (var ev in batch)
            {
                _queue.Enqueue(ev);
            }
            foreach (var ev in pending)
            {
                _queue.Enqueue(ev);
            }
            // Trim if restoring pushed us over the cap.
            while (_queue.Count > _options.MaxQueueSize)
            {
                _queue.Dequeue();
            }
        }
    }

    // Callers must hold _gate.
    private void ArmTimer()
    {
        if (!_disposed)
        {
            _timer.Change(_options.FlushIntervalMs, Timeout.Infinite);
        }
    }

    // Callers must hold _gate.
    private void DisarmTimer() => _timer.Change(Timeout.Infinite, Timeout.Infinite);

    /// <inheritdoc />
    public void Dispose()
    {
        if (_disposed)
        {
            return;
        }
        _disposed = true;
        // Best-effort final flush.
        try
        {
            FlushAsync().GetAwaiter().GetResult();
        }
        catch
        {
            // swallow
        }
        _timer.Dispose();
        if (_ownsHttpClient)
        {
            _httpClient?.Dispose();
        }
    }

    /// <inheritdoc />
    public async ValueTask DisposeAsync()
    {
        if (_disposed)
        {
            return;
        }
        _disposed = true;
        try
        {
            await FlushAsync().ConfigureAwait(false);
        }
        catch
        {
            // swallow
        }
        await _timer.DisposeAsync().ConfigureAwait(false);
        if (_ownsHttpClient)
        {
            _httpClient?.Dispose();
        }
    }
}
